doc/developer/design/20230216_role_based_access_control.md

```diff
@@ -0,0 +1,368 @@
+|---------------------------|-----------------|-------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------|
+| login privilege           | `LOGIN`         | Roles with this attribute can establish a database connection.                                                                                        | Yes             |
+| superuser status          | `SUPERUSER`     | Can bypass all permission checks, except login.                                                                                                       | Yes             |
+| database creation         | `CREATEDB`      | Can create a database.                                                                                                                                | Yes             |
+| role creation             | `CREATEROLE`    | Can create, alter, drop, grant membership to, and revoke membership from roles.                                                                       | Yes             |
+| inheritance of privileges | `INHERIT`       | Can inherit the privileges of roles that it is a member of. On by default. For this project we can keep this as a mandatory attribute.                | Yes             |
+| cluster creation          | `CREATECLUSTER` | Can create a cluster.                                                                                                                                 | No              |
+| persisted data creation   | `CREATEPERSIST` | Can write data to S3, either directly or indirectly (for example inserting into a table would be directly and creating a source would be indirectly). | No              |
+
+These attributes will be added to the attributes accepted by the `CREATE ROLE` statement. Additionally, we will not
+require `LOGIN` or `SUPERUSER` for new roles.
+
+We will add the following SQL statement:
+
+- `ALTER ROLE <role_name> [ WITH ] <option> [ ... ]`
+    - `<role_name>` is the name of the role to alter.
+    - `<option>`: is one of the Option Name’s from above OR one of the option names from above with a `NO` prepended (
+      ex: `NOLOGIN`). An option without a `NO` grants the attribute, an option with a `NO` revokes the attribute. All
+      unmentioned attributes are left unchanged.
+    - Anyone with the `CREATEROLE` attribute can run this on any other role
+        - `SUPERUSER` attribute is required to change the `SUPERUSER` attribute on another role.
+        - `SUPERUSER` can run this without `CREATEROLE`.
+    - `WITH` is ignored.
+
+When a new user logs in, we will create a new role for them with only the `LOGIN` and `INHERIT` attributes. If that user
+is a frontegg admin, then the role will also have the `SUPERUSER` attribute.
```

Comment:
@ggnall Do have an opinion on the following related scenario: A user logs in AND they are a Frontegg admin AND they have an existing role in Materialize AND the existing role is not a super user. Should we update that Materialize role to be a super user or leave it alone?

(FYI @mjibson)

Reply:
Does this scenario play out on every login? Or only on the initial one? 

When we eventually decouple Account permissions (or build our own), it probably makes sense to have SQL/RBAC be the system of record on who can do what. 

From that perspective, an existing explicitly GRANTed role should take precedence. This would enable a case where a user needs to manage users, billing, etc., but should have limited or no database access.

The simplest thing to explain is: "Account admins are always SUPERUSERS", but it seems reasonable to say "Account admins are always SUPERUSERS by default, unless they are explicitly granted another role". 

There will be extra steps in the latter case to upgrade users: 
- OrgAdmin -> OrgMember: would retain explicitly granted roles
- OrgMember -> OrgAdmin: would retain explicitly granted roles. Another SUPERUSER would have to remove existing roles or GRANT them SUPERUSER.

Reply:
From another perspective, I'm not 100% sure how important this scenario is:
>This would enable a case where a user needs to manage users, billing, etc., but should have limited or no database access.

If it's simpler for us to implement "Account admins are always SUPERUSERS" and it's reversible, we could start there.

Reply:
@ggnall The above scenario is about a frontegg user who was initially a non-admin frontegg user, but was then later granted admin in frontegg. Should we bump up their perms in mz? Also applies the other way around: if a frontegg user later gets their admin role revoked, should we also revoke mz superuser?

Reply:
@mjibson We may have crossed messages but for:
> "Account admins are always SUPERUSERS"

...Yes, frontegg admins grants/revokes should probably be paired with mz superuser grants/revokes. It would make sense to keep these in sync, so that users that invite new members to their organization are also able to fully manage access.

Reply:
I'm not sure I fully understand the decision here. I guess the essential question is:

When a Frontegg user connects to Materialize, and we see that the Frontegg user's permissions/privileges don't match the Materialize role's  permissions/privileges, should we update the Materialize role to match the Frontegg user?

It sounds like the answer here is yes, we should update the Materialize role? However, that seems to contradict the statement about SQL/RBAC being the source of truth.

Reply:
Two potential options here are:
1. SQL/RBAC is the system of record. Whenever roles are created/altered in the frontend, SQL statements are sent to Materialize to make these updates to the Materialize roles. The change is considered successful after the SQL statement has executed successfully. In this approach we'd need someway to update the Frontegg roles to match the Materialize roles.
2. Frontegg is the system of record. Whenever a user logs in to Materialize, we update the Materialize role to match the Frontegg role.

I don't know enough about the frontend or Frontegg to know how feasible option 1 is. However, I'm not sure it makes sense to consider SQL/RBAC as the system of record, but also update the SQL/RBAC based on what we see in Frontegg.

Reply:
Yea sorry, I was waxing about both options. I am leaning toward #2 since it's both cleaner to describe to users and doesn't require two-way syncing.

Reply:
I think option 2 makes the most sense right now. It's how our auth is currently set up. Option 1 would likely require much more changes across more teams.

Reply:
Another take: we should disallow `CREATE USER x LOGIN`, and instead derive LOGIN-ability only from frontegg. In addition to syncing superuser status, if a user logs in with valid frontegg creds, they should also always be granted the LOGIN attribute.

Reply:
This would still allow for admins to pre-auth by doing `CREATE ROLE blah ATTRS`, because that user would then later auto get LOGIN (otherwise the CREATE ROLE would have *prevented* their login).

Reply:
I'm OK with syncing both `SUPERUSER` and `LOGIN` attributes every time a user logs in, and for that matter I'd be OK with syncing any arbitrary attribute/privilege/ownership when a user logs in. However, I feel that disallowing `CREATE USER x LOGIN` is unnecessary coupling with Frontegg. It makes it impossible to create new login roles unless Frontegg is being used which will also make it harder for local development and testing.

Another consideration with Frontegg being the source of truth is how do we update roles while they are logged in? If we revoke admin privileges from someone in Frontegg while they are logged in to Materialize, then how does Materialize become aware of this change? We'll likely need the frontend to send an `ALTER ROLE` command.

Reply:
> disallowing CREATE USER x LOGIN is unnecessary coupling with Frontegg

Maybe we should want tight coupling with Frontegg? If it's the source of truth, then we shouldn't let users type things that won't have any effect.

> It makes it impossible to create new login roles unless Frontegg is being used

Frontegg is always being used (except local dev, which I'll discuss below). Let's say we did allow users to specify `LOGIN`. How could any sql driver successfully connect using that role? The *only* way is if they have valid frontegg auth, and which will allow them to login regardless of if the admin typed even `CREATE ROLE x NOLOGIN`. That is: frontegg is the only way connections can authenticate, and authentication also guarantees login-ability. I'm not aware of any use cases for the `LOGIN` keyword that will ever do anything besides confuse people when they think it's doing something it isn't.

> make it harder for local development and testing

We already do other auth/user creation actions if frontegg auth is disabled, so I think we can achieve this. We don't ever need to type `CREATE ROLE x LOGIN` during dev or testing afaik. If we do, it should maybe be removed unless it matches an actual user pattern.

Reply:
Frontegg is always being used now, but I worry that coupling RBAC with Frontegg will lock us into Frontegg and make it challenging to switch if we ever have the need/want to switch. My original hope was that RBAC would be decoupled from Frontegg, be completely implemented in SQL, and any integrations with the frontend would be done via SQL. I can imagine a world where a customer can create a new role on the frontend and a SQL query is immediately sent to Materialize to create the new role, instead of having to wait for that user to login so we can sync their role with Frontegg.

Having said that if the consensus is that we want to couple this feature with Frontegg, then I'm ok with that. Though, I think the next logical question is, if we don't allow specifying the `LOGIN` attribute via SQL, then should we be allowing any RBAC via SQL? We should at least prevent users from typing `SUPERUSER` because that also will be overwritten by Frontegg on login. Though perhaps we should completely remove `CREATE ROLE`, `ALTER ROLE`, `DROP ROLE`, `GRANT`, and `REVOKE`. Instead every time a user logs in we sync all of the user's attributes and privileges with what is in Frontegg. I don't want to end up in a state where the source of truth for some subset of the user's privileges is Frontegg and the source of truth for some other subset of the user's privileges is Materialize.

> We don't ever need to type CREATE ROLE x LOGIN during dev or testing afaik.

In order to test RBAC we will either need to type that or use some other method of creating login roles.

Reply:
> When you log in to a role with Frontegg, you implicitly get `LOGIN` privilege for that session.

What about when you're not using Frontegg? Should we just say that any rolle that is authenticated is implicitly given the `LOGIN` privilege for that session?

Reply:
Yeah, that sounds like the right approach to me. Then in local dev everything implicitly gets `LOGIN`, which sounds about right.

Reply:
> We introduce an mz_has_superuser() function (or maybe there's some existing PostgreSQL function we can piggyback on) that reports whether the current session has SUPERUSER privilege.

PostgreSQL has a read only session variable called `is_superuser` that exposes this, https://www.postgresql.org/docs/current/sql-show.html (grep for "is_superuser"). We have the concept of read only session parameters, but all of them are hard coded into the binary. It should be possible to support this by adding super user status as a param to `SessionVars::new()`.

Reply:
> You're never allowed to specify LOGIN or SUPERUSER when you create a role. The role in Materialize always exists with neither of those characteristics.
...
> When you log in to a role with Frontegg and your JWT has the "Organization Admin" role, you implicitly get SUPERUSER privilege for that session.

One consequence of this is that we lose the mechanism to revoke `SUPERUSER` from someone while they're logged in. So as long as someone remains connected, they retain `SUPERUSER` after it's been revoked in Frontegg. Unless, we invent some new syntax for this and make sure to always update this field in the session before executing a command.

Reply:
> > You're never allowed to specify LOGIN or SUPERUSER when you create a role. The role in Materialize always exists with neither of those characteristics.
> > ...
> > When you log in to a role with Frontegg and your JWT has the "Organization Admin" role, you implicitly get SUPERUSER privilege for that session.
> 
> One consequence of this is that we lose the mechanism to revoke `SUPERUSER` from someone while they're logged in. So as long as someone remains connected, they retain `SUPERUSER` after it's been revoked in Frontegg. Unless, we invent some new syntax for this and make sure to always update this field in the session before executing a command.

Or we have some periodic Frontegg refresh that periodically re-authenticates through Frontegg and updates these attributes.

Reply:
Also I've updated the doc with what we've discussed in this thread.

Reply:
> > Or we have some periodic Frontegg refresh that periodically re-authenticates through Frontegg and updates these attributes.
> 
> We've already got a task that periodically refreshes access tokens for pgwire connections!
> 
> https://github.com/MaterializeInc/materialize/blob/a75724950cf536c5fa22cee625a8e352324fed5a/src/pgwire/src/protocol.rs#L178
> 
> Would be pretty easy to stitch in `SUPERUSER` updates to that task, I think. Would only apply to pgwire sessions and not HTTP sessions, but I think that's ok; pgwire sessions tend to be long lived while http sessions tend to be short lived.

I've updated the doc to include this.